### PR TITLE
Allows for a second migration source format.

### PR DESCRIPTION
**Pull request by brandonparsons**
My attempt to address #1894.

I'm not quite sure about the following line, as I haven't gotten into typespecs at all in Elixir yet:

`@spec run(Ecto.Repo.t, binary | [{integer, module}], atom, Keyword.t) :: [integer]`

### CODE REVIEW

File: lib/ecto/migrator.ex [hunk 1]
```diff
@@ -124,7 +124,14 @@ defmodule Ecto.Migrator do
...
   @doc """
-  Apply migrations in a directory to a repository with given strategy.
+  Apply migrations to a repository with a given strategy.
+
+  The second argument identifies where the migrations are sourced from. A file
+  path may be passed, in which case the migrations will be loaded from this
+  during the migration process. The other option is to pass a list of tuples
+  that identify the version number and migration modules to be run, for example:
+
+  `Ecto.Migrator.run(Repo, [{0, MyApp.Migration1}, {1, MyApp.Migration2}, ...], :up, opts)`
```
Comment (josevalim) [inline lib/ecto/migrator.ex]: You can remove backticks and indent with four spaces.

File: lib/ecto/migrator.ex [hunk 2]
```diff
@@ -170,49 +177,55 @@ defmodule Ecto.Migrator do
...
+  defp migrations_for(migration_source) when is_binary(migration_source) do
+    query = Path.join(migration_source, "*")
 
     for entry <- Path.wildcard(query),
         info = extract_migration_info(entry),
         do: info
   end
 
+  # This function will match specific version/modules passed into `Migrator.run`.
+  defp migrations_for(migration_source) do
```
Comment (josevalim) [inline lib/ecto/migrator.ex]: You can add when `is_list(migration_source)` as a guard then.

File: lib/ecto/migrator.ex [hunk 3]
```diff
@@ -265,10 +273,28 @@ defmodule Ecto.Migrator do
...
+    end
+  end
+
   defp raise_no_migration_in_file(file) do
     raise Ecto.MigrationError,
           "file #{Path.relative_to_cwd(file)} does not contain any Ecto.Migration"
   end
+  defp raise_no_migration_in_module(mod) do
+    raise Ecto.MigrationError,
+          "module #{mod} does not contain any Ecto.Migration"
```
Comment (josevalim) [inline lib/ecto/migrator.ex]: "does not contain any Ecto.Migration" => "is not an Ecto.Migration"

Comment (josevalim) [inline lib/ecto/migrator.ex]: `#{mod}` should be `#{inspect mod}` so we properly format the module name.

File: lib/ecto/migrator.ex (hunk 1 above)
Comment (brandonparsons) [inline lib/ecto/migrator.ex]: ✅

File: lib/ecto/migrator.ex (hunk 2 above)
Comment (brandonparsons) [inline lib/ecto/migrator.ex]: ✅

File: lib/ecto/migrator.ex (hunk 3 above)
Comment (brandonparsons) [inline lib/ecto/migrator.ex]: ✅

**Comment by brandonparsons** [comment]
Thanks for reviewing. Looks like I'm picking up Elixir reasonably well! Addressed your comments in the second commit. I can squash if it all looks good to you.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**